mautic/mautic
# Salesforce latest do not contact

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
When updating a do not contact record check in which integration it was updated last, take the last updated record to sync the systems. At the moment, if you map the do not contact field in Salesforce, this will only update Salesforce based on what Mautic has.

#### Steps to test this PR:
1. Select the option to update Do not contact field by latest updated in the plugin configuration Feature setting tab.
2. Map has opted out of email in integration to Do Not Contact by Email in mautic in the contact fields for either Lead or Contact or both.
2. With this PR we first check where the field was updated last and update the opposite side based on that.
3. In Salesforce make sure the "Email Opt out" field is visible to edit, and that field history tracking has been set for the Email Opt out field.
under the setup menu, select customise Lead and contacts layouts.
<img width="291" alt="screen shot 2017-10-10 at 08 53 28" src="https://user-images.githubusercontent.com/1496976/31375292-f456a3e6-ad98-11e7-8b4b-83d7828bd69b.png">
Add the Email Opt out field to the lead and contact layouts
<img width="333" alt="screen shot 2017-10-10 at 08 53 58" src="https://user-images.githubusercontent.com/1496976/31375293-f46d8d9a-ad98-11e7-9c8f-de42fb584d22.png">
Continue by going to the customise fields
<img width="347" alt="screen shot 2017-10-10 at 08 54 18" src="https://user-images.githubusercontent.com/1496976/31375294-f4847b0e-ad98-11e7-9040-60d823a05ded.png">
select the email opt out field and edit the field level security option
<img width="689" alt="screen shot 2017-10-10 at 08 54 38" src="https://user-images.githubusercontent.com/1496976/31375295-f49bed70-ad98-11e7-8d93-d913e0fbd972.png">
<img width="518" alt="screen shot 2017-10-10 at 08 55 11" src="https://user-images.githubusercontent.com/1496976/31375296-f4c15efc-ad98-11e7-91fa-2b78d4562ec7.png">
Check to see if the field is visible at all levels if its not select it and save
<img width="1015" alt="screen shot 2017-10-10 at 08 55 16" src="https://user-images.githubusercontent.com/1496976/31375297-f4d88c1c-ad98-11e7-8621-f8339d9c7702.png">
<img width="982" alt="screen shot 2017-10-10 at 08 55 27" src="https://user-images.githubusercontent.com/1496976/31375299-f4ee54e8-ad98-11e7-83b6-7679516afcee.png">
Setup Field history by going to the setup->Customize->Leads->Fields menu (also the Contacts fields menu) and set history tracking on the Email Opt Out fields
<img width="685" alt="screen shot 2017-10-11 at 10 30 50" src="https://user-images.githubusercontent.com/1496976/31440552-44f61ba2-ae88-11e7-800e-fb7ea0a3d3dd.png">
<img width="582" alt="screen shot 2017-10-11 at 10 31 04" src="https://user-images.githubusercontent.com/1496976/31440553-450fc2a0-ae88-11e7-9d51-a60f01c6fee9.png">
